engine/crates/runtime-local/src/hooks.rs

```diff
@@ -20,15 +27,89 @@ struct HooksWasiInner {
     gateway: Pool<GatewayComponentInstance>,
     authorization: Pool<AuthorizationComponentInstance>,
     subgraph: Pool<SubgraphComponentInstance>,
+    hook_latencies: Histogram<u64>,
+}
+
+impl HooksWasiInner {
+    async fn run_and_measure<F, T>(&self, hook_name: &'static str, hook: F) -> Result<T, wasi_component_loader::Error>
+    where
+        F: Future<Output = Result<T, wasi_component_loader::Error>>,
+    {
+        let start = SystemTime::now();
+        let result = hook.await;
+        let duration = SystemTime::now().duration_since(start).unwrap_or_default();
+
+        let status = match result {
+            Ok(_) => HookStatus::Success,
+            Err(wasi_component_loader::Error::Internal(_)) => HookStatus::HostError,
+            Err(wasi_component_loader::Error::Guest(_)) => HookStatus::GuestError,
+        };
+
+        let attributes = vec![
```

Comment:
I think you can skip the allocation here, if you're just passing `&attributes` in the next statement.